Add `authctl user set-uid` command

adombeck opened the pull request:
Support setting the UID of a user via `authctl user set-uid <user> <uid>`.

Also changes the owner and group of the user's home directory and all files in the home directory from the old UID and GID to the new UID and GID (if it is owned by the current user), same as `usermod` does when changing the UID of a user.

Closes #630 
UDENG-7717
UDENG-8720

adombeck committed:
Fix comment

There's nothing encrypted in this string.

adombeck committed:
Add gRPC method SetUserID

adombeck committed:
Add `authctl user set-uid` command

adombeck committed:
Return a gRPC error with code PermissionDenied

... instead of prefixing the error message with "permission denied"

adombeck committed:
debian/authd.service.in: Grant capability CAP_DAC_READ_SEARCH

Required to change the ownership of the user's home directory when
changing the user's UID.

adombeck committed:
fileutils: Add ChownRecursiveFrom

We use this to recursively change the owner and group of the user's home
directory when changing the user's UID.

adombeck committed:
Automatically change owner of home directory when changing UID

Do the same usermod does when changing a UID of a user: If the home
directory is currently owned by the user, recursively change the owner
and group of the home directory and all files in the home directory from
the old UID and GID to the new UID and GID.

adombeck commented:
@3v1n0 @denisonbarbosa The PR is still missing tests, but I would appreciate a first review nonetheless, especially regarding the decision to automatically change the owner of the home directory (see e68fd5380d7541ccbc3f71c8e6dc338b31fdd1c1) before I spend time adding tests for that.

denisonbarbosa commented on cmd/authctl/user/set-uid.go:
```diff
@@ -0,0 +1,47 @@
+
+		// Print any warnings returned by the server.
+		for _, warning := range resp.Warnings {
+			fmt.Printf("Warning: %s\n", warning)
```
ditto

adombeck replied:
Those are meant for the user. The warnings we currently use are:

https://github.com/ubuntu/authd/blob/59488fa1dc2a945f3f034b8d4eee0adf6c558f47/internal/users/manager.go#L407-L411

and 

https://github.com/ubuntu/authd/blob/59488fa1dc2a945f3f034b8d4eee0adf6c558f47/internal/users/manager.go#L415-L420

denisonbarbosa commented on internal/services/permissions/permissions.go:
```diff
@@ -41,8 +40,6 @@ func New(args ...Option) Manager {
 // The pid and uid are extracted from peerCredsInfo in the gRPC context.
 func (m Manager) CheckRequestIsFromRoot(ctx context.Context) (err error) {
-	defer decorate.OnError(&err, "permission denied")
-
```
Why this removal?

adombeck committed:
Add TestSetUIDCommand

adombeck committed:
Add db_test.TestSetUserID

adombeck committed:
Support creating users/groups in bubblewrap

We need that for the SetUserID tests

adombeck committed:
testutils: Add RunTestAsRoot

Needed to test fileutils.ChownRecursiveFrom. We can't use bubblewrap for
that because bubblewrap only creates UID mapping for one user, using
chown with a different UID fails with:

    chown: changing ownership of 'file': Invalid argument

adombeck committed:
Add TestChownRecursiveFrom

adombeck committed:
refactor: Split cmd/authctl/user/user_test.go

adombeck commented on internal/users/db/update.go:
```diff
@@ -203,3 +203,75 @@ func (m *Manager) UpdateLockedFieldForUser(username string, locked bool) error {
+	// because UID == GID for user private groups (see https://wiki.debian.org/UserPrivateGroups#UPGs)
+	if _, err := tx.Exec(`UPDATE users SET uid = ?, gid = ? WHERE name = ?`, newUID, newUID, username); err != nil {
+		return err
+	}
```
`authctl user set-uid` now only changes the user's UID. I will add a `authctl group set-gid` command in a follow-up PR.

denisonbarbosa replied:
> I'm considering to deviate from that and make authctl group set-gid also change the group of the home directory of user's whose primary group is updated that way

Although the approach you suggested makes more sense and is more usable, I'm a bit worried about mimicking the behavior of a well known tool but doing it quietly differently. Although, I think we shouldn't have any issues in this case, since we are adding a behavior rather than removing one.

I do prefer the way you suggested, though. I think it's a good one.

adombeck replied:
> Make the behavior clear in the documentation

I added long descriptions for `set-uid`:

```
$ authctl user set-uid --help
Set the UID of a user managed by authd to the specified value.

The new UID value must be unique and non-negative.

The user's home directory and any files within it owned by the user will
automatically have their ownership updated to the new UID. If the home directory
ownership cannot be changed, a warning will be displayed but the command will
still exit successfully.

Files outside the home directory are not updated and must be changed manually 
if needed. To update ownership of all files on the system, use:
  sudo chown -R --from OLD_UID NEW_UID /

This command requires root privileges.

Examples:
  authctl user set-uid john 15000
  authctl user set-uid alice 20000

Usage:
  authctl user set-uid <name> <uid> [flags]

Flags:
  -h, --help   help for set-uid
```

and `set-gid`:

```
$ authctl group set-gid --help
Set the GID of a group managed by authd to the specified value.

The new GID value must be unique and non-negative.

When a group's GID is changed, any users whose primary group is set to this group
will have their primary group GID updated. The home directories of these users and
files within them owned by the group will be updated to the new GID. If changing
ownership fails, a warning will be displayed but the command will still succeed.

Files outside users' home directories are not updated and must be changed manually.
To update group ownership of all files on the system, use:
  sudo chown -R --from :OLD_GID :NEW_GID /

This command requires root privileges.

Examples:
  authctl group set-gid staff 30000
  authctl group set-gid developers 40000

Usage:
  authctl group set-gid <name> <gid> [flags]

Flags:
  -h, --help   help for set-gid
```

adombeck committed:
Add SetGroupID

adombeck committed:
Add db_test.TestSetGroupID

denisonbarbosa commented on internal/users/manager_bwrap_test.go:
```diff
@@ -0,0 +1,132 @@
+		"Successfully_set_UID":                                               {},
+		"Successfully_set_UID_if_ID_is_already_set":                          {uidAlreadySet: true},
+		"Successfully_set_UID_if_ID_is_already_in_use_as_GID_of_system_user": {uidAlreadyInUseAsGIDofSystemUser: true},
+		"Successfully_set_UID_if_ID_is_already_in_use_as_GID_of_authd_user":  {uidAlreadyInUseAsGIDofAuthdUser: true},
```
I don't see an issue with running tests with `sudo` privileges, sometimes we can't avoid it. Just remember that these tests must be skipped when building the package, but you already added helpers for that, so that's fine.

adombeck replied:
damn it, `unshare` doesn't work in the CI. the tests are skipped:

```
    bubblewrap.go:45: Checking if we can use unprivileged user namespaces
    bubblewrap.go:45: Running command: /usr/bin/unshare --user --map-root-user --map-auto /tmp/authd-bwrap-3980848776/bwrap --ro-bind / / --dev /dev --tmpfs /tmp --bind /tmp/TestChownRecursiveFromSuccessfully_change_owner2217086065/001/etc /etc --bind /tmp/go-build1152975855/b473/fileutils.test /tmp/go-build1152975855/b473/fileutils.test --ro-bind /etc/environment /etc/environment --ro-bind /etc/localtime /etc/localtime --ro-bind /etc/login.defs /etc/login.defs --ro-bind /etc/nsswitch.conf /etc/nsswitch.conf --ro-bind /etc/sudo.conf /etc/sudo.conf --ro-bind /etc/sudoers /etc/sudoers --ro-bind-try /etc/timezone /etc/timezone --ro-bind /etc/pam.d /etc/pam.d --ro-bind /etc/security /etc/security --bind /home/runner/work/authd/authd/raw-coverage /home/runner/work/authd/authd/raw-coverage /bin/true
    bwrap: Creating new namespace failed: Operation not permitted
    bubblewrap.go:45: Can't use user namespaces: exit status 1
    bubblewrap.go:52: Running command: /usr/bin/unshare unshare --user --map-root-user --map-auto /tmp/authd-bwrap-3980848776/bwrap --ro-bind / / --dev /dev --tmpfs /tmp --bind /tmp/TestChownRecursiveFromSuccessfully_change_owner2217086065/002/etc /etc --bind /tmp/go-build1152975855/b473/fileutils.test /tmp/go-build1152975855/b473/fileutils.test --ro-bind /etc/environment /etc/environment --ro-bind /etc/localtime /etc/localtime --ro-bind /etc/login.defs /etc/login.defs --ro-bind /etc/nsswitch.conf /etc/nsswitch.conf --ro-bind /etc/sudo.conf /etc/sudo.conf --ro-bind /etc/sudoers /etc/sudoers --ro-bind-try /etc/timezone /etc/timezone --ro-bind /etc/pam.d /etc/pam.d --ro-bind /etc/security /etc/security --bind /home/runner/work/authd/authd/raw-coverage /home/runner/work/authd/authd/raw-coverage /bin/true
    bwrap: Creating new namespace failed: Operation not permitted
    bubblewrap.go:52: Can't use bubblewrap with sudo: exit status 1
    fileutils_test.go:446: Skipping test: requires root privileges or unprivileged user namespaces
```

and in the "Build debian packages" jobs, the `unshare` command hangs forever for some reason.

adombeck replied:
I'm giving up, I can't reproduce the issue locally but `unshare` just doesn't work in the GitHub CI when mapping anything else than the current user's UID. We'll have to execute the tests which require chown as root in the CI.

denisonbarbosa reviewed (commented):
Amazing work. Some small comments, but we should be able to address them quite quickly.

adombeck committed:
tests: Support chown in bubblewrap

adombeck committed:
tests: Run TestChownRecursiveFrom in bubblewrap

We now support chown in bubblewrap, so we don't have to run the test as
root anymore.

adombeck committed:
refactor: Remove unused parameter from runInBubbleWrap

adombeck committed:
testutils/bubblewrap: Support updating golden files

adombeck committed:
testutils: Skip cleanup when in bubblewrap

adombeck committed:
Add users_test.TestSetUserID

adombeck committed:
Add users_test.TestSetGroupID

adombeck committed:
Add `authctl group set-gid` command

adombeck committed:
Add TestSetGIDCommand

adombeck committed:
Add `authctl group` command

adombeck committed:
pam/integration-tests: Don't print authctl usage message in VHS tape

It doesn't test anything that's not already covered by other tests and
it's annoying to have to manually update the golden files of the SSH
integration tests whenever the authctl usage message changes.

adombeck commented:
> Will add the missing tests and also update TestChownRecursiveFrom to use bubblewrap :)

Done and I also added `authctl group set-gid`. Should be ready for another round of review.

adombeck committed:
authctl: Add long description for set-uid command

adombeck committed:
bubblewrap: Mount a tmpfs to /tmp in the sandbox

We have a use case where we want to create a directory at a
deterministic path in /tmp. That fails if /tmp is shared with the host
and other bubblewrap sandboxes which use the same directory.

adombeck committed:
authctl: Add long description for set-gid command

3v1n0 commented on internal/proto/authd/authd.proto:
```diff
@@ -166,6 +168,24 @@ message GetGroupByIDRequest{
+
+message SetGroupIDResponse {
+  repeated string warnings = 1;
+}
```
Given we're likely also adding support for setting the user shell and probably GECOs and maybe in future (with user DB metadata) more fields, isn't better to just have a single `SetUserProperties` message when we handle it all?

As in the daemon side we'll have a single lock place instead of having to manage it for each case we will handle.

adombeck replied:
> As in the daemon side we'll have a single lock place

what do you mean with lock place?

I currently don't see the advantage of a single `SetUserProperties` message. The good thing is that we don't have to be very careful about this API between authd and authctl, because we ship them both in the same package, so we can always change it however we want.

3v1n0 commented on internal/users/manager.go:
```diff
@@ -364,6 +369,178 @@ func compareNewUserInfoWithUserInfoFromDB(newUserInfo, dbUserInfo types.UserInfo
+		warning := fmt.Sprintf("User %q already has UID %d", name, uid)
+		log.Info(context.Background(), warning)
+		return []string{warning}, nil
+	}
```
That said... I'd likely lock the whole thing now when manipulating users through these APIs (including `UpdateUser`).

3v1n0 replied:
> Oh, I thought it would behave exactly like lckpwdf, i.e. block for 15 seconds before returning an error. Is there a reason we made it behave differently?

No, that's the default behavior, since it is meant to give unique access to a process to the locked resources, so if we're owning the lock, we're good to go basically, but we can't lock again (thus the error).

Other processes instead will behave that way.

> We already check that via user.LookupId. We can't use UserDBLocked.IsUniqueUID because that also returns false if the ID is used as a GID.

Well, can't it split it out?

adombeck replied:
> It's a global per-process lock but it would ErrLock in case of concurrent requests rather than blocking

If that's the behavior of `lckpwdf` then the man page should be updated to mention that IMO. Currently it reads like it would always block for 15 seconds:

       The  lckpwdf() function is intended to protect against multiple simultaneous accesses of the shadow password database.
       It tries to acquire a lock, and returns 0 on success, or -1 on failure (lock not obtained within 15 seconds).  The ul‐
       ckpwdf() function releases the lock again.  Note that there is no protection against direct access of the shadow pass‐
       word file.  Only programs that use lckpwdf() will notice the lock.

> Well, can't it split it out?

I don't see why we should change `UserDBLocked`. We don't need it here AFAICT, `userslocking.WriteLock` and `user.LookupId` give us everything we need.

3v1n0 replied:
> userslocking.WriteLock and user.LookupId give us everything we need.

That's fine, but we then need to just not rely on `userslocking.WriteLock` as global lock, or add a version with a mutex on top.

IIRC these were things I somewhat had drafted in the past but then we went for a simpler approach.

> If that's the behavior of lckpwdf then the man page should be updated to mention that IMO

Well, feel free to update libc, but we also are testing this case (in fact we do not hang):

`go test -C internal/users/locking -v -run TestLockAndLockAgainGroupFile`

And it was one of the reasons why I initially added the reflock too.

3v1n0 commented on internal/testutils/bubblewrap.go:
```diff
@@ -70,16 +74,35 @@ func RunTestInBubbleWrap(t *testing.T, args ...string) {
 
-	cmd := exec.Command("bwrap")
+	// Since 25.10 Ubuntu ships the AppArmor profile /etc/apparmor.d/bwrap-userns-restrict
+	// which restricts bwrap and causes chown to fail with "Operation not permitted".
```
Was this reported upstream? Is it something that should not be done?

3v1n0 replied:
Well not here, I'm speaking at wider level in the distro.

3v1n0 commented on internal/testutils/bubblewrap.go:
```diff
@@ -70,16 +74,35 @@ func RunTestInBubbleWrap(t *testing.T, args ...string) {
+	// Since 25.10 Ubuntu ships the AppArmor profile /etc/apparmor.d/bwrap-userns-restrict
+	// which restricts bwrap and causes chown to fail with "Operation not permitted".
+	// We work around that by copying the bwrap binary to a temporary location so that
+	// the AppArmor profile is not applied.
```
Wow... Well definitely a quite relevant apparmor flaw I'd say... Why does it even has apparmor profiles? :)

adombeck replied:
> as otherwise an app could just replicate the same done here

to do what we're doing here an app already needs privileges which allow it to execute arbitrary code as the current process (i.e. the process which launches bwrap). I really don't see an issue here.

3v1n0 commented on internal/users/manager_bwrap_test.go:
```diff
@@ -0,0 +1,419 @@
+	// We use a deterministic path (/tmp/home) here because the home directory
+	// is stored in the database which we dump to a golden file, so we would
+	// have to replace the path in the golden file to make it deterministic.
+	// It's simpler to just use a deterministic path here.
```
This looks like another reason to do something like https://github.com/ubuntu/authd/pull/979/commits/f7c0622b5c5dc1f064c31b4ff7c2b817de6d3095#diff-e8a1ba2baeae8e0a9bceba1f716778bab92beb602eda806dffeddbbd54f2924b as IIRC was done to make possible to make the users home deterministic, not sure if the same applies here though :)

adombeck replied:
I find the solution to use a deterministic path below `/tmp` simple and quite elegant - given that the test is executed with bwrap, so the deterministic path doesn't cause conflicts with other tests.

adombeck replied:
> BTW... Ideally all the integration tests should maybe just go through bubblewrap, so that we can simulate a real scenario better without env variables or preloading hacks.

yes, I also suspect we could simplify things a lot and make tests closer to real scenarios by using bwrap more

adombeck committed:
Take lock in SetUserID/SetGroupID

userslocking.WriteLock() immediately returns ErrLock if the lock is
already taken *by the current process*. lckpwdf behaves similarly (even
though the man page doesn't mention it).

To avoid that issue, we now take another lock which blocks concurrent
goroutines.

adombeck committed:
tests: Avoid unshare hanging forever in some environments

When executing `unshare --map-user` via exec.Command and connecting the
process's stdout or stderr, the command hangs forever if unprivileged
user namespaces are disabled.

We avoid that by checking via `unshare --user` if unprivileged user
namespaces are enabled.

adombeck committed:
tests: Don't skip bubblewrap tests in CI

We broke the bubblewrap tests in the CI without noticing it (at first)
because the tests were skipped. The only case where we really want to
skip the tests is on Launchpad builders. To detect that, we check if the
DEB_BUILD_ARCH environment variable is set and we're *not* in GitHub CI.

adombeck committed:
tests: Make chown work with bubblewrap executed via sudo

adombeck committed:
Try enabling unprivileged user namespaces in CI

adombeck committed:
tests: Skip bubblewrap tests in autopkgtest

The "Run autopkgtests" CI job runs the tests in an LXD container which
doesn't allow using bubblewrap. It fails with:

    bwrap: Failed to make / slave: Permission denied

To avoid that these jobs fail, we allow them to skip the bubblewrap
tests. We still run the tests in the "Go Tests" CI jobs.

adombeck committed:
ci: Group log lines of llvm-symbolizer installation

adombeck committed:
autopkgtest: Run go tests without -v

Running our tests with -v produces so much output that it makes it
harder to inspect test failures, for example when viewing the logs of
the "Run autopkgtests" CI job in GitHub.

Running the tests without -v still prints the logs of the failed tests
which should include all the information we need to debug test failures.

adombeck committed:
authctl: Add argument completion

3v1n0 commented on cmd/authctl/user/set-uid.go:
```diff
@@ -0,0 +1,70 @@
+
+Files outside the home directory are not updated and must be changed manually 
+if needed. To update ownership of all files on the system, use:
+  sudo chown -R --from OLD_UID NEW_UID /
```
And maybe add a warning, even after the operation has completed

3v1n0 replied:
Yeah, that is a quite soft view IMHO, in all the debian packages we never delete users because of this reason (especially the system ones), here the risk is similar IMHO, so I'd follow the guidance of https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1006912 but applied here.

So, I think we should allow to do this, but adding extra warnings to make clear that changing the UID may potentially give access to data to un-authorized parties, isn't wrong.

Note that as per that bug, authd should likely also create fake local users when the package is purged, and it should never remove its own list of owned UIDs/GIDs, but that's another story (for later).

adombeck replied:
I updated the usage message to point out the security risk.

3v1n0 replied:
That's in the help only, right?

I mean, *every time* an UID/GID is changed IMHO we should show that warning too.

adombeck commented on internal/users/manager.go:
```diff
@@ -364,6 +369,184 @@ func compareNewUserInfoWithUserInfoFromDB(newUserInfo, dbUserInfo types.UserInfo
+	log.Debugf(context.Background(), "Changing ownership of home directory %q from UID %d to UID %d", oldUser.Dir, oldUser.UID, uid)
+	err = fileutils.ChownRecursiveFrom(oldUser.Dir, oldUser.UID, 0, int32(uid), -1)
+	if err != nil {
+		warning := "Failed to change ownership of home directory"
```
done

3v1n0 commented on internal/services/user/user.go:
```diff
@@ -218,6 +218,44 @@ func (s Service) ListGroups(ctx context.Context, req *authd.Empty) (*authd.Group
+	if req.GetName() == "" {
+		return nil, status.Error(codes.InvalidArgument, "no user name provided")
+	}
+
```
Oh, yeah it makes sense... It also implies that this change cannot be done if the user session is running (as there will always be a process on such name, unless systemd user session is disabled).

So likely we can just do a simple check on session state first, and then fallback on processes checks.

adombeck replied:
I had to go with the process checks in the end because it's possible that there are processes running as a UID without there being a session for that UID, for example when running a process with `sudo -u <user>`.

adombeck committed:
Check if user is busy before changing its UID

adombeck committed:
testlog: Fix duplicate newlines

adombeck committed:
tests: Improve logging of bubblewrap commands

adombeck committed:
tests: Improve logging when building authd and authctl

adombeck committed:
tests: Improve logging of authctl tests

adombeck committed:
authctl: Tell authd which language to return warnings in

As suggested by reviewer. It's not implemented for now, warnings are
always returned in English.

adombeck committed:
Revert "ci: Install and load apparmor-profiles"

We don't need to load the bwrap-userns-restrict AppArmor profile for the
bubblewrap tests to work. In fact, we even have to circumvent the
AppArmor profile (if it's loaded) for the tests to work.

This reverts commit 7b926c0a669b8473daf4ecb12dfdd3e3eca51783.

aleasto commented:
> So GDM doesn't seem to have a problem with the fact that the UID is outdated in AccountsService. Is there anything else we need to worry about?

The implications may vary depending on which clients are using AccountsService.
In https://bugs.launchpad.net/ubuntu/+source/gdm3/+bug/2134405, gsd-xsettings calling `act_user_set_input_sources` was incorrectly triggering an authentication prompt because it thought the requesting user did not match the target user (uid didn't match).

If `authctl user set-uid` is only to be called while logged-out, and you say that logging-in resolves the discrepancy in AccountsService, then there might be no visible problem.

adombeck committed:
ci/build-deb: Allow the build-deb action to use sudo

Needed to run bubblewrap tests

adombeck committed:
fileutils: Pass structs to ChownRecursiveFrom

adombeck committed:
tests: Pass TESTS_UPDATE_GOLDEN to bubblewrap tests

adombeck committed:
db: Hold mutex when updating the database

adombeck committed:
authctl: Print colored warnings and errors

Uses the same colors as systemctl etc.

adombeck committed:
Use single quotes in user-visible messages

Looks more polished and is consistent with other CLI tools like
systemctl (see for example the output of `systemctl mask foo`).

adombeck committed:
Improve warning messages

adombeck committed:
authctl: Print info messages when UID/GID changed

It's possible that SetUserID returns an error but the UID was still
changed, for example when an error occurred while trying to update the
ownership of the home directory.

In that case, make it clear to the user that the UID was changed.

Also print a message when the home directory was updated.

adombeck committed:
authctl: Print message about manually updating ownership

As requested by reviewer, to make it more visible to the user that they
have to manually update the ownership of files outside the user's home
directory.

adombeck commented on internal/users/db/update.go:
```diff
@@ -193,3 +193,168 @@ func (m *Manager) UpdateLockedFieldForUser(username string, locked bool) error {
+		if _, err := m.db.Exec(`PRAGMA foreign_keys = ON`); err != nil {
+			log.Errorf(context.TODO(), "Failed to re-enable foreign keys: %v", err)
+		}
+	}()
```
done

3v1n0 commented on internal/fileutils/fileutils.go:
```diff
@@ -130,3 +131,48 @@ func LockDir(dir string) (func() error, error) {
+// Symlinks are not followed.
+//
+// If toUID or toGID is negative, the UID or GID is not changed.
+func ChownRecursiveFrom(root string, fromUID, fromGID uint32, toUID, toGID int32) error {
```
Nice!

3v1n0 reviewed (approved):
Just a nit, but LGTM

adombeck committed:
fileutils: Return error in ChownRecursiveFrom if all args are nil

So we don't hide a programmer error.
